### PR TITLE
lr/sc instructions

### PR DESCRIPTION
**Pull request by lvella**
Dismembered from PR #553.

Built upon PRs #615 and #620, because the test file relies on instructions introduced in these PRs.

### CODE REVIEW

File: riscv/src/compiler.rs [hunk 1]
```diff
@@ -1428,6 +1433,40 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
             ]
         }
 
+        insn if insn.starts_with("lr.w") => {
+            // Very similar to "lw":
```
Comment (chriseth) [inline riscv/src/compiler.rs:1437]: Can we somehow add an assertion about what characters come after `"lr.w"` in `insn`?

File: riscv/src/compiler.rs [hunk 2]
```diff
@@ -1428,6 +1433,40 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
...
+            let (rd, rs, off) = rro(args);
+            assert_eq!(off, 0);
+            // TODO misaligned access should raise misaligned address exceptions
+            only_if_no_write_to_zero_vec(
+                vec![
+                    format!("addr <=X= {rs};"),
+                    format!("{rd} <== mload();"),
```
Comment (chriseth) [inline riscv/src/compiler.rs:1444]: The standard says you have to sign-extend the value, or did I read that wrong?

File: riscv/src/compiler.rs [hunk 3]
```diff
@@ -1428,6 +1433,40 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
...
+                    // For consistency sake, I tried to use format!() here, but
+                    // the lint doesn't like it
+                    "lr_sc_reservation <=X= 1;".into(),
```
Comment (chriseth) [inline riscv/src/compiler.rs:1447]: Strictly, we should have one reservation bit per memory address (if I read the standard correctly). If we don't have that, maybe we could scrap the reservation altogether, or would some code break in that case?

Comment (chriseth) [reply riscv/src/compiler.rs:1447]: If we keep this, we should do it even if `rd` is zero.

File: riscv/src/compiler.rs (hunk 2 above)
Comment (lvella) [reply riscv/src/compiler.rs:1444]: I think that part is only relevant for 64-bits when using `lr.w`. On 32 bits, `lr.w` is already the size of the register.

File: riscv/src/compiler.rs (hunk 3 above)
Comment (lvella) [reply riscv/src/compiler.rs:1447]: > Strictly, we should have one reservation bit per memory address (if I read the standard correctly). If we don't have that, maybe we could scrap the reservation altogether, or would some code break in that case?

The reservation can be for any memory size. It seems that a typical implementation is <= 1Kb, but the ISA specs says to assume it to be of any size, so ours is for the entire memory.

> If we keep this, we should do it even if `rd` is zero.

Indeed, makes sense.

File: riscv/src/compiler.rs (hunk 1 above)
Comment (lvella) [reply riscv/src/compiler.rs:1437]: Do you have an idea of where we can find the exhaustive list?

Comment (lvella) [reply riscv/src/compiler.rs:1437]: Found in source: https://github.com/bminor/binutils-gdb/blob/e030ce2c79feee6a35665a6580a23dcf937ea46f/opcodes/riscv-opc.c#L332

I've enumerated all possibilities in "instruction_ends_control_flow", will that be enough?

**Review by chriseth — APPROVED** [review]

**Pull request merged**